### PR TITLE
send updates only when there was a change in the targets.

### PR DESCRIPTION
**Pull request by krasi-georgiev**
this will limit the unnecessary CPU cicles when providers keep sending
updates that didn't actually change anything.

closes: https://github.com/prometheus/prometheus/issues/4518

see https://github.com/kubernetes/client-go/issues/454 for more details.
Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

File: discovery/manager.go [hunk 1]
```diff
@@ -190,7 +186,17 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 			if _, ok := m.targets[poolKey]; !ok {
 				m.targets[poolKey] = make(map[string]*targetgroup.Group)
 			}
-			m.targets[poolKey][tg.Source] = tg
+
+			// No need to update when there aren't any actual changes.
+			// The k8s discoverer keeps sending updates for "kube-controller-manager" and "kube-scheduler" even though these don't actually include any targets.
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply discovery/manager.go]: updated

**Comment by grobie** [comment]
I don't know enough about all the consumer implementations of the manager, and whether some might need such seemingly unnecessary updates to work correctly. Would be good to get another pair of eyes on this @brancz. Also I'd have expected a test to ensure this behavior (duplicates don't get propagated).

**Comment by krasi-georgiev** [comment]
@grobie thanks for the quick feedback. Spot on for both suggestions!

> Instead of adding another channel to the manager struct and increasing the coupling between runProvider and updateGroup functions, I'd have exposed the update status as return value: updateGroup(poolKey poolKey, tgs []*targetgroup.Group) (updated bool).

updated


> Also I'd have expected a test to ensure this behavior (duplicates don't get propagated).

keep forgetting about the tests :stuck_out_tongue_winking_eye:  . Will add one soon.

File: discovery/manager.go [hunk 2]
```diff
@@ -190,9 +191,19 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
...
 				m.targets[poolKey] = make(map[string]*targetgroup.Group)
 			}
-			m.targets[poolKey][tg.Source] = tg
+
+			// No need to update when there aren't any actual changes.
+			// The k8s discoverer keeps sending updates for "kube-controller-manager" and
+			// "kube-scheduler" even though these don't actually include any targets.
+			// Even if we find a way to filter these in the k8s client
+			// this check is safer for other misbehaving providers.
+			if !reflect.DeepEqual(m.targets[poolKey][tg.Source], tg) {
```
Comment (brancz) [inline discovery/manager.go]: `reflect.DeepEqual` seems excessive and maybe even incorrect here, as `DeepEqual` returns false if the ordering of `.Targets` has changed within the `Group`, but that does not actually make a difference. I'd propose to implement an `Equal` method on the `Group`.

As a side note, those events are generated (for the better or worse) because the controller manager and scheduler use `Endpoints` objects to perform leader election. They change an annotation to signify that they are still alive, which results in an even for Prometheus. There's nothing we can do and technically this "works as intended" on the Kubernetes side.

Comment (krasi-georgiev) [reply discovery/manager.go]: I was also crossed for the "DeepEqual" so added an Equal method, but without any sorting for the Targets. Sorting maps would kill all performance gains so the comparison will only work when the targets arrive in the same order. Otherwise we just send a new update.

**Comment by krasi-georgiev** [comment]
@grobie added a new test. Not 100% what I wanted, but a better test would mean somehow exposing the `updateReceived` channel so that the test can listen if it gets notified for a new update.

File: discovery/targetgroup/targetgroup.go [hunk 3]
```diff
@@ -91,3 +91,26 @@ func (tg *Group) UnmarshalJSON(b []byte) error {
 	tg.Labels = g.Labels
 	return nil
 }
+
+// Equal returns true if all fields have exactly the same values and lenghts.
+// To keep a good performance the `Targets []model.LabelSet` is not sorted
+// so if these arrive in a different order the comparing won't work.
```
Comment (brancz) [inline discovery/targetgroup/targetgroup.go]: I think "if .Targets of the compared objects are of different order they are not equal", would be clearer, but I'm also unsure whether this semantic might be a problem, as I don't think we ask provider implementations to have this consistent, just some (like Kubernetes) happen to report targets in a consistent order.

File: discovery/manager.go [hunk 4]
```diff
@@ -190,9 +190,19 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 			if _, ok := m.targets[poolKey]; !ok {
 				m.targets[poolKey] = make(map[string]*targetgroup.Group)
 			}
-			m.targets[poolKey][tg.Source] = tg
+
+			// No need to update when there aren't any actual changes.
+			// The k8s discoverer keeps sending updates for "kube-controller-manager" and
```
Comment (brancz) [inline discovery/manager.go]: We should drop this explanation about a specific target provider and just keep the first line. If we really want to keep the Kubernetes example we should mention why Kubernetes does this.

File: discovery/targetgroup/targetgroup.go (hunk 3 above)
Comment (krasi-georgiev) [reply discovery/targetgroup/targetgroup.go]: another way to  deal with this is to filter these in the k8s provider code, but not sure if another provider might misbehave in the same way.

Comment (brancz) [reply discovery/targetgroup/targetgroup.go]: Filtering them would be incorrect, the events rightfully end up here, it just happens to be a side effect, it could just as well be a valid update.

**Comment by grobie** [comment]
I'm a bit confused by the discussion, if the annotations of the k8s targets change, will this change ever find any duplicate updates? Shouldn't these all differ in their annotation labels?

File: discovery/targetgroup/targetgroup.go (hunk 3 above)
Comment (krasi-georgiev) [reply discovery/targetgroup/targetgroup.go]: `SelfLink:/api/v1/namespaces/kube-system/endpoints/kube-controller-manager`

can this actually include some targets to be scraped? if yes than end of discussion can't filter these.

About the slice order, implementing to order the maps would mean many iterations in a loop so it would be probably worse than just allowing the duplicate updates.

**Comment by krasi-georgiev** [comment]
@grobie  can you include an example as I don't understand the question.

We are discussing the even when receiving 2 identical target groups that have the slice order in a different order we can't catch these as duplicates without ordering the slice(which are maps and sorting these would be too heavy)

https://github.com/krasi-georgiev/prometheus/blob/master/discovery/targetgroup/targetgroup.go#L27

**Comment by grobie** [comment]
My question was more fundamental: do we really receive 2 completely identical target groups? As I understood @brancz explanation above, one annotation label constantly changes, or is this not correct?

**Comment by krasi-georgiev** [comment]
@grobie 

> As a side note, those events are generated (for the better or worse) because the controller manager and scheduler use Endpoints objects to perform leader election. They change an annotation to signify that they are still alive, which results in an even for Prometheus. There's nothing we can do and technically this "works as intended" on the Kubernetes side.

at my observation the only thing that changes is "renewTime" and "ResourceVersion" and we don't care about these so once they hit the Discovery manager any of these are detected as duplicates.

**Comment by brancz** [comment]
Yes and I've also seen those services used to actually discover the controller-manager and scheduler so events from those objects can potentially be real updates, and the resourceVersion changing implies that the apiserver actually accepted a real (non-nop) change, so we can't ignore it.

**Comment by krasi-georgiev** [comment]
@brancz @grobie any final thoughts on this?

**Comment by simonpasquier** [comment]
IMO this PR relies on something (eg sorted labels) that wasn't explicitly expected from the `Discoverer` interface. 

https://github.com/prometheus/prometheus/blob/abf6fe0a983ca116f1c3be8875193458174c6748/discovery/manager.go#L56-L63

It also says `Whenever a discovery provider detects a potential change, it sends the TargetGroup through its channel` which I read like "the manager will forward whatever it receives from the provider".

We could enforce the sorting for all built-in SD providers but now that we encourage people to build their own service discovery, it could be breaking for them.

**Comment by brancz** [comment]
I agree with @simonpasquier. How about instead of checking for equality we hash all the targets with fnv and then hash all those hashes together again, then we have a single comparison value that is not order sensitive.

**Comment by krasi-georgiev** [comment]
after some brainstorming came up with this solution. will run a benchmark to see if it will actually make any performance gains.

**Comment by krasi-georgiev** [comment]
based on the benchmarking I don't see any gains so at this point unless anyone has other ideas I am closing this and can revisit again at some point if needed.
